### ISSUE TITLE
When a block is clicked, why check block id in each block?

### ISSUE THREAD

**Issue by cljoly** [greshake/i3status-rust#249]
I’m wondering why it is up to the `click` function from each block to check [(as advised here)](https://github.com/greshake/i3status-rust/blob/bb49eb575fede955a2facc98b050fe9105b85784/src/block.rs#L18).

Why isn’t this checked [at this point](https://github.com/greshake/i3status-rust/blob/bb49eb575fede955a2facc98b050fe9105b85784/src/main.rs#L241)? This way, each block wouldn’t have to figure out whether it was clicked.

**Comment by atheriel** [comment]
I always thought it was an artifact of the way the i3bar protocol sends events. But it's quite clunky. If you can come up with an alternative I'd happily refactor it.

**Comment by greshake** [comment]
It's because every block can have multiple sub-widgets (buttons, mostly) that can be clicked independently. One example is the music block which has three different buttons:
```rust
fn click(&mut self, event: &I3BarEvent) -> Result<()> {
        if let Some(ref name) = event.name {
            let action = match name as &str {
                "play" => "PlayPause",
                "next" => "Next",
                "prev" => "Previous",
                _ => "",
            };
            if action != "" {
                let m = Message::new_method_call(
                    format!("org.mpris.MediaPlayer2.{}", self.player),
                    "/org/mpris/MediaPlayer2",
                    "org.mpris.MediaPlayer2.Player",
                    action,
                ).block_error("music", "failed to create D-Bus method call")?;
                self.dbus_conn
                    .send(m)
                    .block_error("music", "failed to call method via D-Bus")
                    .map(|_| ())
            } else {
                Ok(())
            }
        } else {
            Ok(())
        }
}
```

**Comment by greshake** [comment]
Also we have the flexibility of allowing blocks to react to arbitrary clicks- I believe this was something I thought was valuable for some advanced concepts I never got to implement (in particular removable storage- imagine a new block popping up for any removable storage with mount/unmount buttons)

**Comment by greshake** [comment]
In a future rewrite, I'd obviously prefer a method where every widget has an implementable click function, but right now that would throw us into some turbulent waters in terms of lifetimes and references.

**Comment by leowzukw** [comment]
Thanks for your answers.

At least, the code to check for the simple case may be factored. If you are ok with that, I would do a PR.

**Comment by greshake** [comment]
We could rename `click` -> `click_any` and add a method `click` that is only called when a widget from the current block is called with an empty default implementation

**Comment by leowzukw** [comment]
Ok, I will have a look at this.

**Comment by MaxVerevkin** [comment]
@ammgws
Should be closed by #1085
